Feature/search posts in helpboard

JayashreeSamantasinghar opened the pull request:
### What does this pull request aim to achieve? 💯

search Icon is visible and clickable in helpboard. After clicking on search unauthorised user can search posts by typing name of other users  

<!--### 🚨Before submitting this pull request🚨:-->

<!--_Please do **NOT** submit this PR if you have not done the following:_-->

- [x] I have checked that no one else is working on similar changes.
- [x] I have read the latest [**README**](README.md) and understand the development workflow.
- [x] The name of this branch is: **`feature/<branch_name>`**, or **`hotfix/<branch_name>`** if this is a hotfix, where `<branch_name>` briefly describes the issue(s) resolved and is prefixed with the issue number(s) (e.g. `1127-1130-update-git-branching-model`). The branch is created in a cloned version of this repo, **not a forked repo**.
- [x] This branch is rebased/merged with the **latest staging** (or the **latest production** if it is a hotfix).
- [x] There are no merge conflicts.
- [x] There are no console warnings and errors.
- [] On the right hand side of this PR, I have linked the appropriate issue(s) under "Issues" and added a project progress under "Projects".
- [x] The title describes the issue(s) being addressed using format: **`<issue_numbers> - <brief_issue_description>`** (e.g. `#1127/#1130 - Update Git Branching Model`)

JayashreeSamantasinghar committed:
search posts in helpboard

JayashreeSamantasinghar committed:
Merge branch 'staging' into feature/search-posts-in-helpboard

JayashreeSamantasinghar committed:
getKeywordSearch

JayashreeSamantasinghar committed:
added testcases

JayashreeSamantasinghar committed:
Merge branch 'staging' into feature/search-posts-in-helpboard

galeza reviewed (changes requested):
Hi @JayashreeSamantasinghar 

I check this test: _Unauthorized user if type single character for search_
First in the POM file I changed:
  _keywordSearchLink = "div[role='button'] input#SCH_IN";
  searchContainerDiv = "div[role='button'] span#SearchContainer div";_
and I added: 
  _getSearchContainerDiv() {
    return cy.get(this.searchContainerDiv);
  }_
#SCH_IN returns two inputs
At the beginning input is not visible, when you click on: searchContainerDiv input expands.

helpBoard.spec.js:

  ```
function searchLinkForPosts() {
    helpBoard.getSearchContainerDiv().click();
    var searchLink = cy.get(helpBoard.getKeywordSearchLink());
    searchLink.should("be.visible").and("have.attr", "type", "text");
    //searchLink.type(feedData.searchBySingleCharacter );
    return searchLink;
  }
```
and in the test:
```
    it.only("Unauthorized user if type single character for search", () => {
      searchLinkForPosts().type(feedData.searchBySingleCharacter + "{enter}");
      cy.get(helpBoard.getErrorMessage()).contains("small", errorMessageText);
    });
```
you don't need click on search icon. you type enter + "{enter}". Cypress type method takes argument:
https://docs.cypress.io/api/commands/type.html#Arguments

galeza commented on cypresse2e/cypress/elements/pages/helpBoard.js:
```diff
@@ -5,6 +5,10 @@ class HelpBoard {
     "div.feed-posts > div:nth-child(1) > div > div > div:nth-child(1)";
   helpBoard = "div.feed-posts";
+  keywordSearchButton = "#SearchContainer > div > img";
+  keywordSearchLink = "#SCH_IN";
```
Link or Input? I would say that you need input

jessica-lau replied:
agree. this should be input.

galeza commented on cypresse2e/cypress/integration/e2eTests/helpBoard/helpBoard.spec.js:
```diff
@@ -1,11 +1,14 @@
   const logo = new Logo();
   var noPostsAvailableText =
-    "Sorry, there are currently no relevant posts available. Please try using a different filter search or";
+    "Sorry, there are currently no relevant posts available. Please try using a different filter search or ";
```
We can move error messages to fixtures. Like for SignIn/JoinIn

galeza commented:
@JayashreeSamantasinghar 
```
    it.only("No post appear if unauthorized user search posts by typing misspelled name", () => {
      searchLinkForPosts().type(feedData.searchByMisspelledText + "{enter}");
      helpBoard.getNoPostsAvailableTextElement().contains(noPostsAvailableText);
    });
```

galeza commented:
```
    it.only("Help Board is empty - Unauthorized user is redirected to SignIn page when clicking Add Post link", () => {
      //there is no post on the Help Board
      searchLinkForPosts().type(feedData.searchByMisspelledText + "{enter}");
      helpBoard.getNoPostsAvailableTextElement().contains(noPostsAvailableText);
      helpBoard.getNoPostsAvailableTextElement().contains(noPostsAvailableText);
      var addPostLink = helpBoard.getAddPostLinkElement();
      addPostLink.click({ force: true });
      cy.validateCorrectScreenIsOpen("auth/login");
    });
  });
```
I repatead code here we need to extract first two lines.

galeza commented on cypresse2e/cypress/fixtures/inputValues.json:
```diff
@@ -1,3 +1,6 @@
 {
-  "location": "New York"
+  "location": "New York",
+  "searchByName": "Sourced by FightPandemics",
```
I don't know how long we are going to use FightPandemics. It will be replaced.

galeza commented:
@JayashreeSamantasinghar 
Hope this helps you. Let me know...

jessica-lau commented on cypresse2e/cypress/integration/e2eTests/helpBoard/helpBoard.spec.js:
```diff
@@ -1,11 +1,14 @@
 import HelpBoard from "../../../elements/pages/helpBoard";
 import Logo from "../../../elements/pages/fpLogo";
+import feedData from "../../../fixtures/inputValues.json";
```
Where is `feedData` coming from? I think it should be `import inputValues from "../../../fixtures/inputValues";` and then you can reference the specific properties in the test cases.

jessica-lau commented on cypresse2e/cypress/integration/e2eTests/helpBoard/helpBoard.spec.js:
```diff
@@ -21,6 +24,54 @@ describe("FightPandemics Help Board Page for unauthorized user", () => {
+        .and("have.attr", "alt", "Icon")
+        .click({ multiple: true, force: true });
+      //cy.keywordSearchIconIsVisibleAndClickable(helpBoard.keywordSearchButton);
+    });
```
Since you already have the searchIcon function written below, you can just simply call `searchIcon().click({ multiple: true, force: true })` in this testcase.

jessica-lau commented on cypresse2e/cypress/elements/pages/helpBoard.js:
```diff
@@ -43,6 +47,22 @@ class HelpBoard {
+  }
+
+  getKeywordSearchLink() {
+    return this.keywordSearchLink;
```
input instead of link

jessica-lau commented on cypresse2e/cypress/integration/e2eTests/helpBoard/helpBoard.spec.js:
```diff
@@ -21,6 +24,54 @@ describe("FightPandemics Help Board Page for unauthorized user", () => {
+      searchIcon().click({ multiple: true, force: true });
+      cy.get(helpBoard.getKeywordSearchLink())
+        .should("be.visible")
+        .and("have.attr", "type", "text");
```
Since you already have the searchLinkForPosts() function written below, you can just simply call `searchLinkForPosts()`

jessica-lau reviewed (changes requested):
I think we also need to test the 'x' in the search as well.

![image](https://user-images.githubusercontent.com/56090896/106218687-cbb3fa00-61a5-11eb-80b0-5c38f24b414b.png)

JayashreeSamantasinghar commented:
> Hi @JayashreeSamantasinghar
> 
> I check this test: _Unauthorized user if type single character for search_
> First in the POM file I changed:
> _keywordSearchLink = "div[role='button'] input#SCH_IN"; searchContainerDiv = "div[role='button'] span#SearchContainer div";_
> and I added:
> _getSearchContainerDiv() { return cy.get(this.searchContainerDiv); }_
> #SCH_IN returns two inputs
> At the beginning input is not visible, when you click on: searchContainerDiv input expands.
> 
> helpBoard.spec.js:
> 
> ```
> function searchLinkForPosts() {
>   helpBoard.getSearchContainerDiv().click();
>   var searchLink = cy.get(helpBoard.getKeywordSearchLink());
>   searchLink.should("be.visible").and("have.attr", "type", "text");
>   //searchLink.type(feedData.searchBySingleCharacter );
>   return searchLink;
> }
> ```
> 
> and in the test:
> 
> ```
>     it.only("Unauthorized user if type single character for search", () => {
>       searchLinkForPosts().type(feedData.searchBySingleCharacter + "{enter}");
>       cy.get(helpBoard.getErrorMessage()).contains("small", errorMessageText);
>     });
> ```
> 
> you don't need click on search icon. you type enter + "{enter}". Cypress type method takes argument:
> https://docs.cypress.io/api/commands/type.html#Arguments

@galeza , I did the changes suggested. All tests passed. Thanks so much for choosing the exact locator names

JayashreeSamantasinghar commented:
> I think we also need to test the 'x' in the search as well.
> 
> ![image](https://user-images.githubusercontent.com/56090896/106218687-cbb3fa00-61a5-11eb-80b0-5c38f24b414b.png)

ok @jessica-lau , I will add test for 'x' in search and will do the changes. Thanks so much for reviewing

JayashreeSamantasinghar commented:
@galeza , @jessica-lau 
I am closing this draft PR because of some merge conflicts and I will create another PR for the same.
